src/exti.rs

```diff
@@ -0,0 +1,135 @@
+//! External interrupt controller (EXTI).
+
+use crate::pac::EXTI;
+
+/// An `ExtiLine` that can be `listen()`ed for interrupt
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct ExtiLine(u8);
+
+/// Internal sources(lines) for ExtiLine
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum InternalLine {
+    Lvd = 16,
+    RtcAlarm = 17,
+    UsbWakeup = 18,
+}
+
+impl ExtiLine {
+    /// Generate an `Option<ExtiLine>` from a GPIO pin configured by Afio
+    ///
+    /// A line from gpio can be obtained through its `pin_number()` method
+    /// ```no_run
+    /// extiline = ExtiLine::from_gpio_line(some_pin.pin_number()).unwrap();
+    /// exti.listen(extiline, TriggerEdge::Falling);
+    /// ```
+    pub fn from_gpio_line(line: u8) -> Option<Self> {
+        match line {
+            0..=15 => Some(ExtiLine(line)),
+            _ => None,
+        }
+    }
+
+    /// Generate an `ExtiLine` from internal source specified by `InternalLine`
+    pub fn from_internal_line(line: InternalLine) -> Self {
+        ExtiLine(line as u8)
+    }
+}
+
+/// Edges that can trigger a configurable interrupt line.
+pub enum TriggerEdge {
+    /// Trigger on rising edges only.
+    Rising,
+    /// Trigger on falling edges only.
+    Falling,
+    /// Trigger on both rising and falling edges.
+    Both,
+}
+
+/// Higher-lever wrapper around the `EXTI` peripheral.
+pub struct Exti {
+    raw: EXTI,
+}
+
+impl Exti {
+    /// Creates a new `Exti` wrapper from the raw `EXTI` peripheral.
+    pub fn new(raw: EXTI) -> Self {
+        Self { raw }
+    }
+
+    /// Destroys this `Exti` instance, returning the raw `EXTI` peripheral.
+    pub fn release(self) -> EXTI {
+        self.raw
+    }
+
+    /// Listen on one of the Lines
+    #[inline]
+    pub fn listen(&mut self, line: ExtiLine, edge: TriggerEdge) {
+        let bm: u32 = 1 << line.0;
+
+        unsafe {
+            match edge {
+                TriggerEdge::Falling => self.raw.ften.modify(|r, w| w.bits(r.bits() | bm)),
+                TriggerEdge::Rising => self.raw.rten.modify(|r, w| w.bits(r.bits() | bm)),
+                TriggerEdge::Both => {
+                    self.raw.ften.modify(|r, w| w.bits(r.bits() | bm));
+                    self.raw.rten.modify(|r, w| w.bits(r.bits() | bm));
+                }
+            }
+
+            self.raw.inten.modify(|r, w| w.bits(r.bits() | bm));
+        }
+    }
+
+    /// Unlisten on the specified line
+    #[inline]
+    pub fn unlisten(&mut self, line: ExtiLine) {
+        let bm: u32 = 1 << line.0;
+
+        unsafe {
+            self.raw.rten.modify(|r, w| w.bits(r.bits() & !bm));
+            self.raw.ften.modify(|r, w| w.bits(r.bits() & !bm));
+            self.raw.inten.modify(|r, w| w.bits(r.bits() & !bm));
+        }
+    }
+
+    /// Enable/Disable event generation on line to wakeup unit
+    #[inline]
+    pub fn gen_event(&mut self, line: ExtiLine, enable: bool) {
```

Comment:
Probably it's better to replace this `boot` with enum

Reply:
Ah yes that's safer, thanks!